### DISCUSSION TITLE
SOCK5 Support

### DISCUSSION THREAD

**Discussion by TuanLeTR** [hierynomus/sshj#841] [Q&A]
Hello All, 
I need to connect to a server via a socks proxy. I am using version 0.35.0. I could not find any method that support socks proxy. Please help. 

Here are the codes that I am using to connect to the server via socks proxy

final SSHClient client = new SSHClient();
try
{
	System.setProperty("java.net.useSystemProxies", "false");
	System.setProperty("socksProxyHost", "my socks proxy server");
	System.setProperty("socksProxyPort", "1080");
	System.setProperty("socksProxyVersion", "5");
	System.setProperty("java.net.socks.username", "");
	System.setProperty("java.net.socks.password", "");
	
	client.addHostKeyVerifier(new PromiscuousVerifier());
	client.connect(my_server_host_name, 2222);
	client.authPassword(my_username, my_password);
}
catch (final UserAuthException | TransportException e)
{
    throw new CobaltException("Error setting client credentials", e);
}

I always got the error message: SOCKS: **Connection not allowed by ruleset.**
Does anyone know how to use SOCK5? Or does this library still support SOCK5? I saw someone add the connect() that supported Proxy back in 2014 but seem like that method was removed.
https://github.com/hierynomus/sshj/pull/141/files

**Comment by hierynomus** [comment]
You can create your own instance of a `javax.net.SocketFactory` that creates a SOCKS Proxy socket. If you pass the SocketFactory to SSHClient using the `SSHClient.setSocketFactory(...)` method, it will connect through SOCKS

**Reply by TuanLeTR** [reply]
Can you please provide an example of what you meant?

**Reply by hierynomus** [reply]
Create a class like:

```
package net.schmizz.sshj;

import java.io.IOException;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Proxy;
import java.net.Socket;
import java.net.UnknownHostException;

import javax.net.SocketFactory;

class ProxySocketFactory extends SocketFactory {

    private Proxy proxy;

    public ProxySocketFactory(InetSocketAddress proxyAddress) {
        this.proxy = new Proxy(Proxy.Type.SOCKS, proxyAddress);
    }

    @Override
    public Socket createSocket(InetAddress address, int port, InetAddress localAddress, int localPort)
            throws IOException {
        Socket s = new Socket(proxy);
        s.bind(new InetSocketAddress(localAddress, localPort));
        s.connect(new InetSocketAddress(address, port));
        return s;
    }

    @Override
    public Socket createSocket(InetAddress host, int port) throws IOException {
        Socket s = new Socket(proxy);
        s.connect(new InetSocketAddress(host, port));
        return s;
    }

    @Override
    public Socket createSocket(String host, int port) throws IOException, UnknownHostException {
        Socket s = new Socket(proxy);
        s.connect(new InetSocketAddress(host, port));
        return s;
    }

    @Override
    public Socket createSocket(String host, int port, InetAddress localHost, int localPort)
            throws IOException, UnknownHostException {
        Socket s = new Socket(proxy);
        s.bind(new InetSocketAddress(localHost, localPort));
        s.connect(new InetSocketAddress(host, port));
        return s;
    }
}
```

And pass that to the `SSHClient.setSocketFactory(...)` before you call `connect(...)`

**Reply by TuanLeTR** [reply]
I got different error using the new class. Here is my new codes.

client.addHostKeyVerifier(new PromiscuousVerifier());
final InetSocketAddress proxyAddr = new InetSocketAddress(proxyHost, proxyPort);
final ProxySocketFactory socketFactory = new ProxySocketFactory(proxyAddr);
client.setSocketFactory(socketFactory);
client.connect(server, port);

I think when I call the client.connect(), this is where the error happens.

java.net.SocketException: Unconnected sockets not implemented

Do you have any thought on what am I missing here? Thanks

**Reply by TuanLeTR** [reply]
I found out the reason I got the error: java.net.SocketException: Unconnected sockets not implemented because in the new ProxySocketFactory class, I didn't have the constructor with no argument.

 I added new override method in the new class

@Override
		public Socket createSocket() 
		{
	        return new Socket();
	    }
This method is called by the connect(String hostname, int port) in the SocketClient class in the package net.schmizz.sshj. After that, still I got the same error as before: SOCKS: Connection not allowed by ruleset

When I debugged throug the codes, these settings are set for the SOCK5
System.setProperty("java.net.useSystemProxies", "false");
System.setProperty("socksProxyHost", "my socks proxy server");
System.setProperty("socksProxyPort", "1080");
System.setProperty("socksProxyVersion", "5");
System.setProperty("java.net.socks.username", "");
System.setProperty("java.net.socks.password", "");

without these settings (even with using the new proxy factory class), the codes throw errors complaining about the server that I am trying to connect doesn't exist (since that server is only available through socks proxy)
 
Any thought on how to resolve this would be much appreciated.

**Comment by exceptionfactory** [comment]
The [Connection is not allowed by ruleset](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/net/SocksSocketImpl.java#L484) message is a response from the SOCKS proxy server indicating that the connection to the specified SSH host and port is not permitted. The error implies that the SOCKS server configuration needs to be changed in order to permit the client to make the SSH connection.

**Reply by TuanLeTR** [reply]
Thanks for the response. This might be the issue then. One question: in my code before calling the client.connect(), I have this line
client.addHostKeyVerifier(new PromiscuousVerifier());

Do you know if I need to add the hostname and the public key to the .ssh/known_hosts file in my windows box? or is this truly something that the proxy server/socks server needs to change some configurations in order for this to work?

Thanks

**Reply by exceptionfactory** [reply]
You're welcome!

The `PromiscuousVerifier` accepts and allows connections to all remote SSH hosts. The `PromiscuousVerifier` is good for testing, and is not related to the proxy connection itself. Adding the SSH public key of the remote server to a `known_hosts` file should be considered after getting the initial connection working.

**Comment by Edwardiv1** [comment]
That's a lot of work to simply specify a proxy. Is there no way to pass in a java.net.Proxy class?

**Reply by exceptionfactory** [reply]
@Edwardiv1 I wrote and released a library called [Socket Broker](https://github.com/exceptionfactory/socketbroker) to provide a configurable Socket Factory that supports either SOCKS5 or HTTP proxy access, with or without authentication. The library has no external dependencies, and is available from Maven Central.

The following post describes the design, and includes example integration steps.

https://exceptionfactory.com/posts/2022/01/24/introducing-socket-broker/

The [BrokeredSocketFactory](https://javadoc.io/doc/com.exceptionfactory.socketbroker/socketbroker/latest/com/exceptionfactory/socketbroker/BrokeredSocketFactory.html) can be plugged in to the SSH Client using `setSocketFactory()` as described in earlier comments.
